Implement test framework for shortest path calculation

hawkinsw opened the pull request:
Closes #7.

hawkinsw committed:
Implement test framework for shortest path calculation

Closes #7.

karlobermeyer commented on README.md:
```diff
@@ -127,7 +127,23 @@ Description: a subroutine of `visibiity_polygon_demonstration.m` and
+calculations.
 
+### Visibility polygon calculation testing
+TBD
```
Good idea to do this in stages with the shortest path test having priority.

hawkinsw replied:
I'm glad that you are on the same page. I was worried. Thanks for the confirmation!

karlobermeyer commented on README.md:
```diff
@@ -127,7 +127,23 @@ Description: a subroutine of `visibiity_polygon_demonstration.m` and
+
+Every test requires three files:
+1. `<name>.environment`: The environment in which to find a shortest path between two points. Specify this like any other environment file.
+2. `<name>.guards`: The location of 2 (two) points within in the environment and not in a hole. The first point is the start point and the second point is the end point. The test will calculate the shortest path between those points with respect to the environment defined in `<name>.environment`.
```
"points within in the" --> "points in the"

hawkinsw replied:
Good :eye:

karlobermeyer commented on README.md:
```diff
@@ -127,7 +127,23 @@ Description: a subroutine of `visibiity_polygon_demonstration.m` and
+
+### Shortest path calculation testing
+
+The script `run_sp_.sh` will execute all the tests in the `sp_tests` directory.
```
There are many things "sp" could stand for, so I recommend using "shortest_path" explicitly rather than the abbreviation "sp", i.e., rename
"run_sp_.sh" --> "run_shortest_path_tests.sh" (or maybe just "test_shortest_path.sh"?)
"sp_tests" --> "shortest_path_tests" etc.

hawkinsw replied:
I actually agree. I was being too flippant. Names are changed!

karlobermeyer commented on src/run_sp_tests.sh:
```diff
@@ -0,0 +1,59 @@
+reset_format='\033[0m'
+
+# Specify that all the shortest path tests are in the sp_tests directory.
+sp_tests_dir="sp_tests"
```
Better to be explicit.
I think it's Ok to use the "sp" abbreviation local to this script because the narrower context makes the meaning clear, so
sp_tests_dir="shortest_path_tests"

hawkinsw replied:
Let me know what you think about the updated code. I _think_ that I understood what you are asking and made the appropriate updates.

karlobermeyer commented on src/visilibity.hpp:
```diff
@@ -2051,6 +2060,71 @@ class Visibility_Graph {
 std::ostream &operator<<(std::ostream &outs,
                          const Visibility_Graph &visibility_graph);
 
```
Doesn't have to be part of this PR, but these days I consider the superior multi-line style to be
```
std::ostream &operator<<(
  std::ostream &outs,
  const Visibility_Graph &visibility_graph,
);
```
I find that easier to read, plus it allows names on the first line to be changed without re-indentation of all subsequent lines. Can that be set in the autoformatter? If not, we don't have to worry about it.

hawkinsw replied:
I will look into this. I am not really too familiar with the autoformatter, TBH. Also, I think that git/github are just showing this here for context because it's not a function that I added/rewrote in this PR. 

That said, I will definitely go back and look at the autoformatter configuration and see what can be done. Is that okay?

karlobermeyer replied:
If you figure it out, super. Or, if you are too busy and don't get to it, that's Ok too.

hawkinsw replied:
I am going to merge this PR now but I am going to continue to fiddle with configuration for clang-format. I think that there are a few options that I can already see might be useful but getting them in the right combination is going to be tricky. I will open a separate issue for handling this, if that's okay?

hawkinsw replied:
If you agree, just go ahead and resolve. Otherwise, comment away :-)

karlobermeyer reviewed (changes requested):
Looks great! This will let us more quickly merge changes with confidence. I left only a few minor naming and style comments.

hawkinsw commented:
I updated the code in the PR -- feel free to comment on the comments that I left and my attempts at addressing your suggestions. Thanks for the thorough feedback!

Will

karlobermeyer reviewed (approved):
LGTM

karlobermeyer commented:
Merge at will.
